### ISSUE TITLE
thriftrw integration

### ISSUE THREAD

**Issue by abhinav** [uber/tchannel-python#149]
This issue is mostly to document/discuss how we want the `thriftrw` integration to look. One of the prerequisites is that we don't want to break the existing support for Apache Thrift-generated code for at least one minor release.

### Current API

Suppose we have a service `KeyValue` defined in `service.thrift`.

* Server: Both, old and new server APIs expect to get a reference to the `service.KeyValue` module generated by Thrift. With that and the name of the method being implemented, they're able to determine enough information about the request and response structure.
* Client: Both `client_for` and `thrift_request_builder` expect a reference to the `service.KeyValue` module. They both iterate through the functions defined for the service and generate methods for the returned objects that know how to make requests for those functions.

### Proposed changes

The user will call `thriftrw` directly and get the compiled module like so,

    service = thriftrw.load('service.thrift')

And pass in the generated service class where they would pass in the service module but only in the new APIs.

    @tchannel.thrift.register(service.KeyValue)
    def foo():
         # ...

    KeyValue = thrift_request_builder('service', service.KeyValue)

    def bar():
        tchannel.thrift(KeyValue.getItem(..))

This is different from what we initially discussed where the user would not interact with `thriftrw` directly but pass in the path to the Thrift file and we'd wrap `thriftrw` in TChannel. That's because originally the plan was for them to pass in the path to the Thrift file only to `thrift_request_builder` but that does nothing for the server-side. We need the same information on the server side as well and this looks like the cleanest way to make that happen.

I'm open to suggestions of alternatives to this.

---

CC @blampe @breerly @junchaowu @kriskowal

**Comment by blampe** [comment]
What about
```
service = tchannel.thrift.install('hyperbahn-name-or-hostport', 'service.thrift')
```
then `service.KeyValue` is still what we register against, and `service.KeyValue.Foo()` is wrapped to return `ThriftRequest`s?

**Comment by abhinav** [comment]
And `service.Bar` is still how you access types defined in `service.thrift`? Also, does this mean that users have to pass references to `service` around instead of being able to import from it anywhere in their code base? It does have the advantage of making sure that people keep all code for talking to a specific external service in one place.

**Comment by junchaowu** [comment]
tchannel.thrift.install looks good.

**Comment by breerly** [comment]
`tchannel.thrift.install` doesn't make any sense because it would just pass through to `thrift_request_builder`...

I think we can streamline things a bit more and do:

```
from tchannel import TChannel, thrift_request_builder

tchannel = TChannel('my-service')

vehicles_service = thrift_request_builder('vehicles', 'vehicles.thrift')
puppies_service = thrift_request_builder('puppies', 'puppies.thrift')
``

And then elsewhere in their app:

```
from app.tchannel import tchannel, vehicles_service

resp = yeild tchannel.thrift(vehicles_service.fetch_all())
```

**Comment by abhinav** [comment]
We discussed this in person and settled on:

```python
from tchannel import TChannel
from tchannel import thrift

donuts = thrift.load('donuts', 'donuts.thrift')
DonutService = donuts.DonutService

# server:

app = TChannel()

@app.register(DonutService)
def deliverDonuts(request):
    order = request.body.order
    # ...

# client:

client = TChannel()
order = donuts.Order(...)
clent.thrift(DonutService.deliverDonuts(order))
```

So, the `tchannel.thrift` module will offer the `load` function.

**Comment by breerly** [comment]
:+1: 

Let's make `thrift_request_builder` throw warnings for the one minor release it still exists.

**Comment by abhinav** [comment]
Addendum: Clients can do,

    thriftrw.load('donuts', 'donuts.thrift', hostport=foo)

For when they want to specify a specific hostport to which the requests will be forwarded.

**Comment by blampe** [comment]
Cool.

**Comment by abhinav** [comment]
Resolved by #161.
